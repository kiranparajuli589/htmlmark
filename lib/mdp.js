import { Lexer } from "./lexer/index.js"
import { Parser } from "./parser/index.js"

/**
 * Markdown Parser for Javascript Application
 */
export class MDP {
	indent
	codeHighlightFn
	constructor(config = {}) {
		this.indent = config.indent || 4
		this.codeHighlightFn = config.codeHighlightFn || null
	}
	/**
	 * Runs the HTML parsing for the given lines of text
	 *
	 * @param {Array} lines - An array of lines of text
	 *
	 * @returns {string} - The parsed HTML
	 */
	h(lines) {
		const lexer = new Lexer(lines)
		const parser = new Parser(lexer.run(), null, this.codeHighlightFn)
		return parser.run()
	}

	/**
	 * Runs HTML parsing for the lines of text
	 *
	 * @param {Array} lines - An array of lines of text
	 *
	 * @returns {string} - The parsed HTML
	 */
	hT(lines) {
		const start = Date.now()
		const html = this.h(lines)
		const end = Date.now()
		console.info(`\x1b[35mParsing time: ${end - start}ms\x1b[0m`)
		return html
	}

	/**
	 * For demo preview
	 *
	 * @param {String} str - The string to parse
	 * @param {Function|null} codeHighlightFn - A function to highlight code blocks
	 *
	 * @returns {{ elapsedTime: number, html: string, lex: Object}}
	 */
	hP(str, codeHighlightFn = null) {
		const lines = str.split("\n")
		const start = Date.now()
		const lexer = new Lexer(lines)
		const lex = lexer.run()
		const parser = new Parser(lex, null, codeHighlightFn)
		const html = parser.run()
		const end = Date.now()
		return {
			elapsedTime: end - start,
			html,
			lex
		}
	}
}
