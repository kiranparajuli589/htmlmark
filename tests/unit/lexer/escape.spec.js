import { MDP } from "../../mdp.js"


const mdp = new MDP()

describe("escape", () => {
	it("should escape >, <, &, ', \" characters", () => {
		const lines = [
			">",
			"<",
			"&",
			"'",
			"\"",
			"<>&'\""
		]
		expect(mdp.h(lines)).toMatchSnapshot()
	})
	it("should escape inside block elements", () => {
		const lines = [
			"# <>&'\""
		]
		expect(mdp.h(lines)).toMatchSnapshot()
	})
	it("should not escape unescaped characters", () => {
		const lines = [
			"\\# not a heading",
			"\\> not a blockquote",
			"\\* not a list",
			"\\` not code `",
			"\\_ not italic_",
			"\\*\\* not bold**",
			"\\~ not strikethrough~",
			"\\[some](http://example.com) not a link",
			"\\![some](http://example.com) not an image",
			"\\| not a table |"
		]
		expect(mdp.h(lines)).toMatchSnapshot()
	})
})
