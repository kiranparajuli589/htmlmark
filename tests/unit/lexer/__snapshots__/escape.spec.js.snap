// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`escape should escape >, <, &, ', " characters 1`] = `
"<p>&gt; &lt; &amp; &#39; &quot; &lt;&gt;&amp;&#39;&quot;</p>
"
`;

exports[`escape should escape inside block elements 1`] = `
"<h1>&lt;&gt;&amp;&#39;&quot;</h1>
"
`;

exports[`escape should not escape unescaped characters 1`] = `
"<p># not a heading \\\\&gt; not a blockquote * not a list \` not code \` _ not italic_ ** not bold** ~ not strikethrough~ [some](http://example.com) not a link !<a href=\\"http://example.com\\">some</a> not an image | not a table |</p>
"
`;
